File: backend/src/ws/messages.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use crate::ws::hub::QuizPhase;

/// Message types for game WebSocket
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum GameMessage {
    #[serde(rename = "join")]
    Join {
        user_id: Uuid,
        session_code: String,
    },
    #[serde(rename = "answer")]
    Answer {
        question_id: Uuid,
        selected_answer: String,
        response_time_ms: i32,
    },
    #[serde(rename = "start_game")]
    StartGame,
    #[serde(rename = "next_question")]
    NextQuestion,
    #[serde(rename = "reveal_answer")]
    RevealAnswer,
    #[serde(rename = "show_leaderboard")]
    ShowLeaderboard,
    #[serde(rename = "end_game")]
    EndGame,
    #[serde(rename = "pass_presenter")]
    PassPresenter {
        next_presenter_user_id: Uuid,
    },
}

/// Server-sent messages
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    #[serde(rename = "connected")]
    Connected { participants: Vec<ParticipantMessage> },
    #[serde(rename = "participant_joined")]
    ParticipantJoined { user: ParticipantMessage },
    #[serde(rename = "participant_left")]
    ParticipantLeft { user_id: Uuid },
    #[serde(rename = "game_started")]
    GameStarted,
    #[serde(rename = "question")]
    Question {
        question_id: Uuid,
        question_number: i32,
        total_questions: i32,
        text: String,
        answers: Vec<String>,
        time_limit: i32,
    },
    #[serde(rename = "time_update")]
    TimeUpdate { remaining_seconds: i32 },
    #[serde(rename = "answer_received")]
    AnswerReceived { user_id: Uuid },
    #[serde(rename = "reveal")]
    Reveal {
        question_id: Uuid,
        question_number: i32,
        question_text: String,
        correct_answer: String,
        distribution: Vec<AnswerDistributionMessage>,
        segment_leaderboard: Vec<LeaderboardEntry>,
        event_leaderboard: Vec<LeaderboardEntry>,
    },
    #[serde(rename = "scores_update")]
    ScoresUpdate {
        scores: Vec<ScoreUpdate>,
    },
    #[serde(rename = "leaderboard")]
    Leaderboard {
        rankings: Vec<LeaderboardEntry>,
    },
    #[serde(rename = "game_ended")]
    GameEnded,
    #[serde(rename = "error")]
    Error { message: String },
    #[serde(rename = "processing_status")]
    ProcessingStatus {
        step: String,           // "transcribing", "generating", "ready"
        progress: Option<i32>,  // 0-100 percentage
        message: String,
    },
    #[serde(rename = "display_mode")]
    DisplayMode {
        mode: String,           // "leaderboard", "question_results", "quiz_progress", "final_results"
        data: serde_json::Value,
    },
    #[serde(rename = "phase_changed")]
    PhaseChanged {
        phase: QuizPhase,
        question_index: i32,
        total_questions: i32,
    },
    #[serde(rename = "all_answered")]
    AllAnswered {
        answer_count: usize,
        total_participants: usize,
    },
    #[serde(rename = "presenter_changed")]
    PresenterChanged {
        previous_presenter_id: Uuid,
        new_presenter_id: Uuid,
        new_presenter_name: String,
        segment_id: Uuid,
    },
    #[serde(rename = "segment_complete")]
    SegmentComplete {
        segment_id: Uuid,
        segment_title: String,
        presenter_name: String,
        segment_leaderboard: Vec<LeaderboardEntry>,
        event_leaderboard: Vec<LeaderboardEntry>,
        segment_winner: Option<LeaderboardEntry>,
        event_leader: Option<LeaderboardEntry>,
    },
    #[serde(rename = "event_complete")]
    EventComplete {
        event_id: Uuid,
        final_leaderboard: Vec<LeaderboardEntry>,
        winner: Option<LeaderboardEntry>,
        segment_winners: Vec<SegmentWinner>,
    },
}

/// Segment winner information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SegmentWinner {
    pub segment_id: Uuid,
    pub segment_title: String,
    pub winner_name: String,
    pub winner_score: i32,
}

/// Participant info in messages
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParticipantMessage {
    pub id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
}

/// Answer distribution in reveal message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnswerDistributionMessage {
    pub answer: String,
    pub count: i32,
    pub is_correct: bool,
}

/// Score update
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoreUpdate {
    pub user_id: Uuid,
    pub username: String,
    pub score: i32,
    pub delta: i32,
}

/// Leaderboard entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LeaderboardEntry {
    pub rank: i32,
    pub user_id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
    pub score: i32,
}

/// Audio WebSocket messages
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AudioMessage {
    #[serde(rename = "audio_chunk")]
    AudioChunk { data: String, timestamp: i64 }, // base64 encoded
    #[serde(rename = "audio_stop")]
    AudioStop,
}

/// Audio server messages
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AudioServerMessage {
    #[serde(rename = "transcript_update")]
    TranscriptUpdate { text: String, is_final: bool },
    #[serde(rename = "question_generated")]
    QuestionGenerated {
        question: String,
        correct_answer: String,
        source_transcript: String,
    },
    #[serde(rename = "transcription_error")]
    TranscriptionError { error: String },
}

/// Canvas WebSocket messages
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum CanvasMessage {
    #[serde(rename = "draw_stroke")]
    DrawStroke {
        stroke: StrokeData,
    },
    #[serde(rename = "clear_canvas")]
    ClearCanvas,
}

/// Canvas server messages
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum CanvasServerMessage {
    #[serde(rename = "stroke_added")]
    StrokeAdded {
        user_id: Uuid,
        username: String,
        stroke: StrokeData,
    },
    #[serde(rename = "canvas_cleared")]
    CanvasCleared,
    #[serde(rename = "canvas_sync")]
    CanvasSync {
        strokes: Vec<StrokeData>,
    },
}

/// Stroke data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StrokeData {
    pub points: Vec<Point>,
    pub color: String,
    pub width: f64,
}

/// Point in a stroke
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

/// Flappy Bird input messages from clients
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum FlappyInputMessage {
    #[serde(rename = "flap")]
    Flap {
        user_id: Uuid,
    },
}

/// Per-player Flappy Bird state
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlappyPlayerState {
    pub user_id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
    pub y: f32,
    pub velocity: f32,
    pub alive: bool,
    pub score: i32,
}

/// Flappy Bird game state broadcast from server
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum FlappyServerMessage {
    #[serde(rename = "flappy_state")]
    FlappyState {
        players: Vec<FlappyPlayerState>,
        obstacle_x: f32,
        gap_y: f32,
    },
}

File: backend-python/app/ws/messages.py
"""WebSocket message type definitions."""

from datetime import datetime
from enum import Enum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class QuizPhase(str, Enum):
    """Quiz phase states."""

    NOT_STARTED = "not_started"
    SHOWING_QUESTION = "showing_question"
    REVEALING_ANSWER = "revealing_answer"
    SHOWING_LEADERBOARD = "showing_leaderboard"
    BETWEEN_QUESTIONS = "between_questions"
    SEGMENT_COMPLETE = "segment_complete"
    MEGA_QUIZ_READY = "mega_quiz_ready"
    MEGA_QUIZ = "mega_quiz"
    EVENT_COMPLETE = "event_complete"
    PRESENTER_PAUSED = "presenter_paused"


# Client -> Server messages
class JoinMessage(BaseModel):
    type: str = "join"
    user_id: UUID
    session_code: str


class AnswerMessage(BaseModel):
    type: str = "answer"
    question_id: UUID
    selected_answer: str
    response_time_ms: int


class StartGameMessage(BaseModel):
    type: str = "start_game"


class NextQuestionMessage(BaseModel):
    type: str = "next_question"


class RevealAnswerMessage(BaseModel):
    type: str = "reveal_answer"


class ShowLeaderboardMessage(BaseModel):
    type: str = "show_leaderboard"


class EndGameMessage(BaseModel):
    type: str = "end_game"


class PassPresenterMessage(BaseModel):
    type: str = "pass_presenter"
    next_presenter_user_id: UUID


class StartMegaQuizMessage(BaseModel):
    type: str = "start_mega_quiz"
    question_count: int | None = None


class SkipMegaQuizMessage(BaseModel):
    type: str = "skip_mega_quiz"


class AdminSelectPresenterMessage(BaseModel):
    type: str = "admin_select_presenter"
    presenter_user_id: UUID
    segment_id: UUID


class SelectPresenterMessage(BaseModel):
    """Host or current presenter selects the next presenter."""
    type: str = "select_presenter"
    presenter_user_id: UUID


class StartPresentationMessage(BaseModel):
    """Presenter starts their presentation (creates segment + starts recording)."""
    type: str = "start_presentation"
    title: str | None = None


class ResumeSegmentMessage(BaseModel):
    """Resume an existing segment."""
    type: str = "resume_segment"
    segment_id: UUID


# Server -> Client messages
class ParticipantInfo(BaseModel):
    user_id: UUID
    username: str
    avatar_url: str | None = None
    join_status: str | None = None
    is_late_joiner: bool = False
    joined_at: datetime | None = None
    online: bool = True


class ConnectedMessage(BaseModel):
    type: str = "connected"
    participants: list[ParticipantInfo]


class ParticipantJoinedMessage(BaseModel):
    type: str = "participant_joined"
    user: ParticipantInfo


class ParticipantLeftMessage(BaseModel):
    type: str = "participant_left"
    user_id: UUID
    online: bool = False


class GameStartedMessage(BaseModel):
    type: str = "game_started"


class GameEndedMessage(BaseModel):
    type: str = "game_ended"


class QuestionMessage(BaseModel):
    type: str = "question"
    question_id: UUID
    question_number: int
    total_questions: int
    text: str
    answers: list[str]
    time_limit: int


class TimeUpdateMessage(BaseModel):
    type: str = "time_update"
    remaining_seconds: int


class AnswerReceivedMessage(BaseModel):
    type: str = "answer_received"
    user_id: UUID


class AnswerDistribution(BaseModel):
    answer: str
    count: int
    percentage: float


class RevealMessage(BaseModel):
    type: str = "reveal"
    question_id: UUID
    question_number: int
    question_text: str
    correct_answer: str
    distribution: list[AnswerDistribution]
    segment_leaderboard: list[dict[str, Any]]
    event_leaderboard: list[dict[str, Any]]


class LeaderboardMessage(BaseModel):
    type: str = "leaderboard"
    rankings: list[dict[str, Any]]


class PhaseChangedMessage(BaseModel):
    type: str = "phase_changed"
    phase: QuizPhase
    question_index: int
    total_questions: int


class AllAnsweredMessage(BaseModel):
    type: str = "all_answered"
    answer_count: int
    total_participants: int


class ErrorMessage(BaseModel):
    type: str = "error"
    message: str


class MegaQuizReadyMessage(BaseModel):
    type: str = "mega_quiz_ready"
    event_id: UUID
    available_questions: int
    current_leaderboard: list[dict[str, Any]]
    is_single_segment: bool = False
    single_segment_mode: Literal["remix", "skip"] | None = None


class MegaQuizStartedMessage(BaseModel):
    type: str = "mega_quiz_started"
    event_id: UUID
    question_count: int


class SegmentCompleteMessage(BaseModel):
    type: str = "segment_complete"
    segment_id: UUID
    segment_title: str
    presenter_name: str
    segment_leaderboard: list[dict[str, Any]]
    event_leaderboard: list[dict[str, Any]]
    segment_winner: dict[str, Any] | None = None
    event_leader: dict[str, Any] | None = None


class SegmentWinner(BaseModel):
    segment_id: UUID
    segment_title: str
    winner_name: str
    winner_score: int


class EventCompleteMessage(BaseModel):
    type: str = "event_complete"
    event_id: UUID
    final_leaderboard: list[dict[str, Any]]
    winner: dict[str, Any] | None = None
    segment_winners: list[SegmentWinner] = Field(default_factory=list)


class PresenterChangedMessage(BaseModel):
    type: str = "presenter_changed"
    previous_presenter_id: UUID
    new_presenter_id: UUID
    new_presenter_name: str
    segment_id: UUID


class PresenterDisconnectedMessage(BaseModel):
    type: str = "presenter_disconnected"
    presenter_id: UUID
    presenter_name: str
    segment_id: UUID


class PresenterPausedMessage(BaseModel):
    type: str = "presenter_paused"
    presenter_id: UUID
    presenter_name: str
    segment_id: UUID
    question_index: int
    total_questions: int
    reason: str | None = None


class PresenterOverrideNeededMessage(BaseModel):
    type: str = "presenter_override_needed"
    presenter_id: UUID
    presenter_name: str
    segment_id: UUID


class NoQuestionsGeneratedMessage(BaseModel):
    type: str = "no_questions_generated"
    segment_id: UUID
    segment_title: str | None = None
    presenter_name: str
    reason: str = "insufficient_content"


class ParticipantNameChangedMessage(BaseModel):
    type: str = "participant_name_changed"
    user_id: UUID
    old_name: str
    new_name: str


class JoinLockStatusChangedMessage(BaseModel):
    type: str = "join_lock_status_changed"
    event_id: UUID
    join_locked: bool
    locked_at: datetime | None = None
    message: str


class QuizGeneratingMessage(BaseModel):
    """Notify clients quiz generation started."""
    type: Literal["quiz_generating"] = "quiz_generating"
    segment_id: UUID


class QuizReadyMessage(BaseModel):
    """Notify clients quiz is ready."""
    type: Literal["quiz_ready"] = "quiz_ready"
    segment_id: UUID
    questions_count: int
    auto_start: bool = True


class ProcessingStatusMessage(BaseModel):
    """Real-time processing status for host."""
    type: Literal["processing_status"] = "processing_status"
    segment_id: UUID
    stage: str
    message: str


class StateRestoredMessage(BaseModel):
    type: str = "state_restored"
    event_id: UUID
    segment_id: UUID | None
    current_phase: QuizPhase
    current_question_id: UUID | None = None
    question_text: str | None = None
    answers: list[str] = Field(default_factory=list)
    time_limit: int | None = None
    question_started_at: datetime | None = None
    your_score: int = 0
    your_answer: str | None = None
    participants: list[dict[str, Any]] = Field(default_factory=list)


class PongMessage(BaseModel):
    type: str = "pong"


class PresenterSelectedMessage(BaseModel):
    """Notify all clients that a presenter has been selected."""
    type: str = "presenter_selected"
    presenter_id: UUID
    presenter_name: str
    is_first_presenter: bool = False


class PresentationStartedMessage(BaseModel):
    """Notify all clients that a presentation has started."""
    type: str = "presentation_started"
    segment_id: UUID
    presenter_id: UUID
    presenter_name: str


class WaitingForPresenterMessage(BaseModel):
    """Notify participants they are waiting for presenter selection."""
    type: str = "waiting_for_presenter"
    event_id: UUID
    participant_count: int


def parse_client_message(data: dict[str, Any]) -> BaseModel | None:
    """Parse a client message based on its type."""
    msg_type = data.get("type")
    parsers = {
        "join": JoinMessage,
        "answer": AnswerMessage,
        "start_game": StartGameMessage,
        "next_question": NextQuestionMessage,
        "reveal_answer": RevealAnswerMessage,
        "show_leaderboard": ShowLeaderboardMessage,
        "end_game": EndGameMessage,
        "pass_presenter": PassPresenterMessage,
        "admin_select_presenter": AdminSelectPresenterMessage,
        "start_mega_quiz": StartMegaQuizMessage,
        "skip_mega_quiz": SkipMegaQuizMessage,
        "select_presenter": SelectPresenterMessage,
        "start_presentation": StartPresentationMessage,
        "resume_segment": ResumeSegmentMessage,
        "pong": PongMessage,
    }
    parser = parsers.get(msg_type)
    if parser:
        return parser.model_validate(data)
    return None
